File: code_samples/section12/example_2_binary_search/binary_search.cs
// ================================
// Original binary search
// ================================

/**
 * Performs a standard binary search on a sorted integer array.
 *
 * @param arr     Sorted array to search
 * @param target  Value to locate
 *
 * @return Index of the target if found; otherwise -1
 */
static int BinarySearch(int[] arr, int target)
{
    int lo = 0;                 // Lower bound of the search range (inclusive)
    int hi = arr.Length - 1;    // Upper bound of the search range (inclusive)

    // Continue searching while the range is valid
    while (lo <= hi)
    {
        // Compute midpoint safely to avoid integer overflow
        int mid = lo + (hi - lo) / 2;

        // Check if the midpoint element matches the target
        if (arr[mid] == target)
        {
            return mid;         // Target found
        }
        // Target is larger; discard left half
        else if (arr[mid] < target)
        {
            lo = mid + 1;
        }
        // Target is smaller; discard right half
        else
        {
            hi = mid - 1;
        }
    }

    // Target was not found in the array
    return -1;
}

// Same, but counts the number of loop iterations (steps)

/**
 * Performs a binary search while counting how many iterations occur.
 *
 * @param arr     Sorted array to search
 * @param target  Value to locate
 *
 * @return Tuple containing:
 *         - index: position of the target, or -1 if not found
 *         - steps: number of loop iterations performed
 */
static (int index, int steps) BinarySearchSteps(int[] arr, int target)
{
    int lo = 0;                 // Lower bound of search range
    int hi = arr.Length - 1;    // Upper bound of search range
    int steps = 0;              // Counts number of iterations performed

    // Continue searching while the range is valid
    while (lo <= hi)
    {
        steps++;  // Count this iteration / comparison step

        // Compute midpoint of the current search range
        int mid = lo + (hi - lo) / 2;

        // Check if midpoint matches the target
        if (arr[mid] == target)
            return (mid, steps);

        // Decide which half of the array to search next
        if (arr[mid] < target)
            lo = mid + 1;       // Target lies in the right half
        else
            hi = mid - 1;       // Target lies in the left half
    }

    // Target was not found after exhausting search range
    return (-1, steps);
}

// ================================
// Load ordered.txt using CURRENT DIRECTORY
// ================================

/**
 * Loads the ordered integer dataset from disk.
 *
 * The file path is resolved relative to the current working directory,
 * which depends on where the script is executed from.
 *
 * @return Array of integers read from ordered.txt,
 *         or an empty array on failure
 */
static int[] LoadOrderedFile()
{
    // When you run:
    //   cd code_samples\section12\lesson_3_binary_search
    //   dotnet script binary_search.csx
    //
    // Environment.CurrentDirectory will be:
    //   code_samples\section12\lesson_3_binary_search
    //
    // The data file lives at:
    //   ..\data\ordered.txt
    string relativePath = @"..\data\ordered.txt";

    // Resolve relative path into a full absolute path
    string fullPath = Path.GetFullPath(relativePath, Environment.CurrentDirectory);

    Console.WriteLine($"Attempting to read: {fullPath}");

    try
    {
        // Read entire file contents as a single string
        string text = File.ReadAllText(fullPath);

        // Split on any whitespace, parse tokens as integers,
        // and convert the result to an array
        return [.. text
            .Split((char[])null, StringSplitOptions.RemoveEmptyEntries)
            .Select(int.Parse)];
    }
    catch (Exception ex)
    {
        // Provide diagnostic output to help troubleshoot file issues
        Console.WriteLine("Error reading ordered.txt");
        Console.WriteLine(ex.Message);
        Console.WriteLine($"Working directory: {Environment.CurrentDirectory}");

        // Return empty array to signal failure
        return [];
    }
}

// ================================
// MAIN (top-level for dotnet script)
// ================================

// Load the ordered dataset from disk
var arr = LoadOrderedFile();

// Validate successful file load
if (arr.Length == 0)
{
    Console.WriteLine("Missing input file - aborting.");
    return; // Exit script early on failure
}

Console.WriteLine("=== Binary Search Tests ===");
Console.WriteLine($"Loaded {arr.Length} integers\n");

// Test 1: Search for the first element (best-case scenario)
var r = BinarySearchSteps(arr, arr[0]);
Console.WriteLine($"Search first  ({arr[0]}): index={r.index}, steps={r.steps}");

// Test 2: Search for the middle element
int midVal = arr[arr.Length / 2];
r = BinarySearchSteps(arr, midVal);
Console.WriteLine($"Search middle ({midVal}): index={r.index}, steps={r.steps}");

// Test 3: Search for the last element (worst-case successful search)
int lastVal = arr[arr.Length - 1];
r = BinarySearchSteps(arr, lastVal);
Console.WriteLine($"Search last   ({lastVal}): index={r.index}, steps={r.steps}");

// Test 4: Search for a value guaranteed not to exist
r = BinarySearchSteps(arr, 999_999);
Console.WriteLine($"Search missing (999999): index={r.index}, steps={r.steps}");

File: code_samples/section12/lesson_3_binary_search/binary_search.java
package code_samples.section12.lesson_3_binary_search;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class binary_search {

    // ----------------------------------------------------
    // Given binary search
    // ----------------------------------------------------
    int binarySearch(int[] arr, int target) {
        int lo = 0, hi = arr.length - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (arr[mid] == target) {
                return mid;
            } else if (arr[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        return -1;
    }

    // Same algorithm, but counts how many iterations (steps)
    int binarySearchSteps(int[] arr, int target, int[] stepsOut) {
        int lo = 0, hi = arr.length - 1;
        int steps = 0;

        while (lo <= hi) {
            steps++;
            int mid = lo + (hi - lo) / 2;
            if (arr[mid] == target) {
                stepsOut[0] = steps;
                return mid;
            } else if (arr[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        stepsOut[0] = steps;
        return -1;
    }

    // ----------------------------------------------------
    // Load integers from file, using project root (user.dir)
    // ----------------------------------------------------
    static int[] loadOrderedFromProjectRoot() {
        String root = System.getProperty("user.dir");
        // This assumes your project structure:
        // <root>/code_samples/section12/data/ordered.txt
        File file = new File(root
                + File.separator + "code_samples"
                + File.separator + "section12"
                + File.separator + "data"
                + File.separator + "ordered.txt");

        System.out.println("Attempting to read: " + file.getAbsolutePath());

        List<Integer> list = new ArrayList<>();
        try (Scanner sc = new Scanner(file)) {
            while (sc.hasNextInt()) {
                list.add(sc.nextInt());
            }
        } catch (FileNotFoundException e) {
            System.out.println("Error reading ordered.txt");
            e.printStackTrace();
            return new int[0];
        }

        int[] arr = new int[list.size()];
        for (int i = 0; i < list.size(); i++) {
            arr[i] = list.get(i);
        }
        return arr;
    }

    // ----------------------------------------------------
    // MAIN TESTS
    // ----------------------------------------------------
    public static void main(String[] args) {
        int[] arr = loadOrderedFromProjectRoot();
        if (arr.length == 0) {
            System.out.println("Missing or empty ordered.txt, aborting.");
            return;
        }

        System.out.println("=== Binary Search Tests ===");
        System.out.println("Loaded " + arr.length + " integers\n");

        binary_search bs = new binary_search();
        int[] steps = new int[1];

        // First element
        int idx = bs.binarySearchSteps(arr, arr[0], steps);
        System.out.println("Search first  (" + arr[0] + "): index=" + idx + ", steps=" + steps[0]);

        // Middle element
        int midVal = arr[arr.length / 2];
        idx = bs.binarySearchSteps(arr, midVal, steps);
        System.out.println("Search middle (" + midVal + "): index=" + idx + ", steps=" + steps[0]);

        // Last element
        int lastVal = arr[arr.length - 1];
        idx = bs.binarySearchSteps(arr, lastVal, steps);
        System.out.println("Search last   (" + lastVal + "): index=" + idx + ", steps=" + steps[0]);

        // Missing element
        idx = bs.binarySearchSteps(arr, 999_999, steps);
        System.out.println("Search missing (999999): index=" + idx + ", steps=" + steps[0]);
    }
}
